lib/tools/apk-utils.js

```diff
@@ -198,6 +198,47 @@ apkUtilsMethods.install = async function (apk, replace = true, timeout = 60000)
   }
 };
 
+apkUtilsMethods.installOrUpgrade = async function (apk, pkg = null, timeout = 60000) {
+  let apkInfo = null;
+  if (!pkg) {
+    apkInfo = await this.getApkInfo(apk);
+    pkg = apkInfo.name;
+  }
+  if (!pkg) {
+    log.warn(`Cannot read the package name of ${apk}. Assuming correct app version is already installed`);
+    return;
+  }
+  if (!await this.isAppInstalled(pkg)) {
+    await this.adbExec(['install', apk], {timeout});
+    return;
+  }
+  const pkgInfo = this.getPackageInfo(pkg);
+  const pkgVersionCode = pkgInfo.versionCode;
+  if (!apkInfo) {
```

Comment:
Has anything changed by this point in terms of what `this.getApkInfo(apk)` will return? If not, why not just immediately do
```js
let apkInfo = await this.getApkInfo(apk);
pkg = pkg || apkInfo.name;
```